packages/fx-core/tests/core/hooks.test.ts

```diff
@@ -875,4 +891,183 @@ describe("Middleware", () => {
       assert(res2.isErr() && res2.error.name === "EmptySelectOption");
     });
   });
+
+  describe("ProjectUpgraderMW", () => {
+    const sandbox = sinon.createSandbox();
+    const appName = randomAppName();
+    const projectSettings = MockProjectSettings(appName);
+    let envJson: Json = {};
+    let userData: Record<string, string> = {};
+
+    const inputs: Inputs = { platform: Platform.VSCode };
+    inputs.projectPath = path.join(os.tmpdir(), appName);
+    const envName = projectSettings.currentEnv;
+    const confFolderPath = path.resolve(inputs.projectPath, `.${ConfigFolderName}`);
+    const settingsFile = path.resolve(confFolderPath, "settings.json");
+    const envJsonFile = path.resolve(confFolderPath, `env.${envName}.json`);
+    const userDataFile = path.resolve(confFolderPath, `${envName}.userdata`);
+
+    function MockFunctions() {
+      sandbox.stub<any, any>(fs, "readJson").callsFake(async (file: string) => {
+        if (settingsFile === file) return projectSettings;
+        if (envJsonFile === file) return envJson;
+        return {};
+      });
+      sandbox.stub<any, any>(fs, "writeFile").callsFake(async (file: string, content: any) => {
+        if (userDataFile === file) {
+          userData = deserializeDict(content);
+        }
+        if (envJsonFile === file) {
+          envJson = JSON.parse(content);
+        }
+      });
+      sandbox.stub<any, any>(fs, "readFile").callsFake(async (file: string) => {
+        if (userDataFile === file) return serializeDict(userData);
+        return {};
+      });
+    }
+
+    beforeEach(() => {
+      sandbox.stub<any, any>(fs, "pathExists").callsFake(async (file: string) => {
+        if (userDataFile === file) return true;
+        if (inputs.projectPath === file) return true;
+        return {};
+      });
+    });
+
+    afterEach(() => {
+      sandbox.restore();
+    });
+
+    it("Previous context and userdata", async () => {
+      envJson = MockPreviousContext();
+      userData = MockPreviousUserData();
+      MockFunctions();
+
+      class MyClass {
```

Comment:
Can we give a meaningful class name and name property? Such as UpgradeHook or something else?

Reply:
Fixed. Thanks!